common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ErrorHandlerSuite.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.network.shuffle;
+
+import java.net.ConnectException;
+
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+/**
+ * Test suite for {@link ErrorHandler}
+ */
+public class ErrorHandlerSuite {
+
+  @Test
+  public void testPushErrorRetry() {
+    ErrorHandler.BlockPushErrorHandler handler = new ErrorHandler.BlockPushErrorHandler();
+    assertFalse(handler.shouldRetryError(new RuntimeException(new IllegalArgumentException(
+      ErrorHandler.BlockPushErrorHandler.TOO_LATE_MESSAGE_SUFFIX))));
+    assertFalse(handler.shouldRetryError(new RuntimeException(new ConnectException())));
+    assertTrue(handler.shouldRetryError(new RuntimeException(new IllegalArgumentException(
+      ErrorHandler.BlockPushErrorHandler.BLOCK_APPEND_COLLISION_DETECTED_MSG_PREFIX))));
```

Comment:
I'm wondering why this error is retrying. My understanding is that the error returns when there's already another attempt wrote or being writing (not sure if this case also included) the same shuffle block. So does retry is for the case where the writing attempt fails to write the block completely?

Reply:
Ah..sorry. I do mean the same shuffle partition rather than the same block.

WDYT?

Reply:
It's retriable because this block hasn't been appended to the merged shuffle file and the merge operation hasn't been finalized yet.

Reply:
When we append a block to the merged shuffle file, we either append it completely or we end up effectively writing nothing to the file.
If the first attempt failed because of collision, then that block effectively hasn't been appended to the file yet, which makes it retriable.
In the 2nd PR to be sent out soon, it will include more details for this part.

Reply:
> If the first attempt failed because of collision, then that block effectively hasn't been appended to the file yet, which makes it retriable.

What makes the first attempt failed because of collision? With my understanding, it has two possibilities:

1. the same partition has been already merged by another task attempt
2. the same partition is merging by another task attempt


For case 1, do we still need to retry? If we do retry in this case, doesn't it return `BLOCK_APPEND_COLLISION_DETECTED_MSG_PREFIX` again? 

For case2, I think it may make sense to retry in case of that attempt doesn't merge partition successfully at the end.

Reply:
We actually distinguish between a block duplication and a block collision on the server side.

Block duplication is when the exact same shuffle partition block gets pushed by different executors, due to speculation or maybe task retry.
The server side is able to tell when block duplication happens, whether is one duplicate block sent after the first has been successfully merged or when both blocks are received at the same time.
With duplicate block, the server will actually respond success to the client, so the client won't retry sending it.
In the case of speculation, when 2 clients might be sending the same block at the same time, the server will respond success to 1 of the two and let the other write, and if that write fails the corresponding client will retry if it's retriable.

On the other hand, a block collision is not about the exact same shuffle partition block, but 2 different blocks belonging to the same shuffle partition being sent to the same shuffle service at the same time.
Since the shuffle service need to append 1 block completely before appending the content of the next block belonging to the same shuffle partition, when these blocks arrive at one shuffle service at the same time, we would encounter a block collision.
A block collision might not  immediately lead to the collision failure sent back to the client, since the server will buffer the blocks for a short period of time and make a few attempts before giving up.
When the shuffle service gives up, the client will receive the collision failure.
If it receives the collision failure, it's an indication that this block hasn't been merged yet, and thus it's OK to retry.
Of course, it's totally possible that by the time the retry happens, a speculative task has already pushed the block and successfully merged it.
If that's the case, the retry would be treated as a block duplication instead of a block collision, and the client will receive success response.

I hope this servers as an overview of what's to come in the next PR.

Reply:
I see. Thanks for the detailed explanation.